crates/apollo-mcp-server/src/introspection/tools/search.rs

```diff
@@ -0,0 +1,207 @@
+use tracing::debug;
+
+/// The name of the tool to search a GraphQL schema.
+pub const SEARCH_TOOL_NAME: &str = "search";
+
+/// The depth of nested types to include for leaf nodes on matching root paths.
+pub const LEAF_DEPTH: DepthLimit = DepthLimit::Limited(1);
+
+/// A tool to search a GraphQL schema.
+#[derive(Clone)]
+pub struct Search {
+    schema: Arc<Mutex<Valid<Schema>>>,
+    index: SchemaIndex,
+    allow_mutations: bool,
+    pub tool: Tool,
+}
+
+/// Input for the search tool.
+#[derive(JsonSchema, Deserialize)]
+pub struct Input {
+    /// The search terms
+    terms: Vec<String>,
+}
+
+/// An error while indexing the GraphQL schema.
+#[derive(Debug, thiserror::Error)]
+pub enum IndexingError {
+    #[error("Unable to index schema: {0}")]
+    IndexingError(#[from] apollo_schema_index::error::IndexingError),
+
+    #[error("Unable to lock schema: {0}")]
+    TryLockError(#[from] tokio::sync::TryLockError),
+}
+
+impl Search {
+    pub fn new(
+        schema: Arc<Mutex<Valid<Schema>>>,
+        allow_mutations: bool,
+    ) -> Result<Self, IndexingError> {
+        let root_types = if allow_mutations {
+            OperationType::Query | OperationType::Mutation
+        } else {
+            OperationType::Query.into()
+        };
+        let locked = &schema.try_lock()?;
+        Ok(Self {
+            schema: schema.clone(),
+            index: SchemaIndex::new(locked, root_types)?,
+            allow_mutations,
+            tool: Tool::new(
+                SEARCH_TOOL_NAME,
+                "Search a GraphQL schema",
+                schema_from_type!(Input),
+            ),
+        })
+    }
+
+    pub async fn execute(&self, input: Input) -> Result<CallToolResult, McpError> {
+        let mut root_paths = self
+            .index
+            .search(input.terms.clone(), Options::default())
+            .map_err(|e| {
+                McpError::new(
+                    ErrorCode::INTERNAL_ERROR,
+                    format!("Failed to search index: {e}"),
+                    None,
+                )
+            })?;
+
+        root_paths.truncate(5);
+        debug!(
+            "Root paths for search terms: {}\n{}",
+            input.terms.join(", "),
+            root_paths
+                .iter()
+                .map(ToString::to_string)
+                .collect::<Vec<String>>()
+                .join("\n"),
+        );
+
+        let schema = self.schema.lock().await;
+        let mut tree_shaker = SchemaTreeShaker::new(&schema);
+        for root_path in root_paths {
+            let types = root_path.inner.types.clone();
+            let path_len = types.len();
+            for (i, type_name) in types.into_iter().enumerate() {
+                if let Some(extended_type) = schema.types.get(type_name.as_ref()) {
+                    let depth = if i == path_len - 1 {
+                        LEAF_DEPTH
+                    } else {
+                        DepthLimit::Limited(1)
+                    };
```

Comment:
I'm a bit confused here. Do both branches result in exactly the same value since `LEAF_DEPTH` is set to `DepthLimit::Limited(1)` on L24?